cri/stream/runtime.go

```diff
@@ -0,0 +1,162 @@
+package stream
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os/exec"
+	"strings"
+	"time"
+
+	apitypes "github.com/alibaba/pouch/apis/types"
+	"github.com/alibaba/pouch/cri/stream/remotecommand"
+	"github.com/alibaba/pouch/daemon/mgr"
+
+	"github.com/sirupsen/logrus"
+)
+
+// Runtime is the interface to execute the commands and provide the streams.
+type Runtime interface {
+	// Exec executes the command in pod.
+	Exec(ctx context.Context, containerID string, cmd []string, streamOpts *remotecommand.Options, streams *remotecommand.Streams) (uint32, error)
+
+	// Attach attaches to pod.
+	Attach(ctx context.Context, containerID string, streamOpts *remotecommand.Options, streams *remotecommand.Streams) error
+
+	// PortForward forward port to pod.
+	PortForward(ctx context.Context, name string, port int32, stream io.ReadWriteCloser) error
+}
+
+type streamRuntime struct {
+	containerMgr mgr.ContainerMgr
+}
+
+// NewStreamRuntime creates a brand new stream runtime.
+func NewStreamRuntime(ctrMgr mgr.ContainerMgr) Runtime {
+	return &streamRuntime{containerMgr: ctrMgr}
+}
+
+// Exec executes a command inside the container.
+func (s *streamRuntime) Exec(ctx context.Context, containerID string, cmd []string, streamOpts *remotecommand.Options, streams *remotecommand.Streams) (uint32, error) {
+	createConfig := &apitypes.ExecCreateConfig{
+		Cmd:          cmd,
+		AttachStdin:  streamOpts.Stdin,
+		AttachStdout: streamOpts.Stdout,
+		AttachStderr: streamOpts.Stderr,
+		Tty:          streamOpts.TTY,
+	}
+
+	execid, err := s.containerMgr.CreateExec(ctx, containerID, createConfig)
+	if err != nil {
+		return 0, fmt.Errorf("failed to create exec for container %q: %v", containerID, err)
+	}
+
+	attachConfig := &mgr.AttachConfig{
+		Streams:     streams,
+		MuxDisabled: true,
+	}
+
+	err = s.containerMgr.StartExec(ctx, execid, attachConfig)
+	if err != nil {
+		return 0, fmt.Errorf("failed to start exec for container %q: %v", containerID, err)
+	}
+
+	// TODO Find a better way instead of the dead loop
+	var ei *apitypes.ContainerExecInspect
+	for {
+		ei, err = s.containerMgr.InspectExec(ctx, execid)
+		if err != nil {
+			return 0, fmt.Errorf("failed to inspect exec for container %q: %v", containerID, err)
+		}
+		// Loop until exec finished.
+		if !ei.Running {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	return uint32(ei.ExitCode), nil
+}
+
+// Attach attaches to a running container.
+func (s *streamRuntime) Attach(ctx context.Context, containerID string, streamOpts *remotecommand.Options, streams *remotecommand.Streams) error {
+	attachConfig := &mgr.AttachConfig{
+		Stdin:   streamOpts.Stdin,
+		Stdout:  streamOpts.Stdout,
+		Stderr:  streamOpts.Stderr,
+		Streams: streams,
+	}
+
+	err := s.containerMgr.Attach(ctx, containerID, attachConfig)
+	if err != nil {
+		return fmt.Errorf("failed to attach to container %q: %v", containerID, err)
+	}
+
+	<-streams.StreamCh
```

Comment:
The channel is just used for the `notify` event?

Reply:
cc @YaoZengzeng

Reply:
@fuweid Yes, it is used to notify the container io has been closed and then the `Attach` function could return, the http connection could be closed.

Reply:
roger